vulture_os/toolkit/api_parser/perception_point_x_ray/perception_point_x_ray.py

```diff
@@ -0,0 +1,237 @@
+            if last_scans.get('count') >= 1000:
+                to = to - (to - since) / 2
+                msg = f"API returns more than 1000 logs. Updating 'to' to {to} for log_type {log_type}"
+                logger.info(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
+            else:
+                break
+
+        logs = []
+        for scan in last_scans.get('results', []):
+            scan_log = self.get_scan_details(scan["full_scan_id"])
+            logs.append(scan_log)
+
+        return logs, last_scans.get('has_more') == True, last_scans.get('next'), to
+    @staticmethod
+    def format_log(log):
+        log['timestamp_epoch'] = log['timestamp']
+        # Remove unused and big fields
+        unused_fields = ["timestamp", "search_descendants", "warning_texts", "scan_tree", "warning_texts", "similarity_content_vector", "disclaimers"]
+        for field in unused_fields:
+            try:
+                del log[field]
+            except KeyError:
+                pass
+        return json.dumps(log)
+
+    def execute(self):
+        # Warning : the fetched logs are ordered in ASC (no option is available in API doc)
+
+        for log_type in self.LOG_TYPES:
+            since = self.last_collected_timestamps.get(f"perception_point_x_ray_{log_type}") or (timezone.now() - timedelta(days=30))
+            to = min(since + timedelta(days=1), timezone.now())
+
+            msg = f"Parser starting from {since} to {to} for log_type {log_type}"
+            logger.info(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
+
+            has_more_logs = True
+            next_url = None
+            while not self.evt_stop.is_set() and has_more_logs:
+
+                logs, has_more_logs, next_url, to = self.get_logs(since, to, log_type, next_url)
+
+                self.write_to_file([self.format_log(log) for log in logs])
+                # Downloading may take some while, so refresh token in Redis
+                self.update_lock()
+
+                if logs and has_more_logs:
+                    max_timestamp = max(int(log.get("timestamp_epoch") * 10000000) for log in logs)
+                    last_timestamp = datetime.fromtimestamp(max_timestamp/10000000, tz=timezone.utc)
+                    self.last_collected_timestamps[f"perception_point_x_ray_{log_type}"] = last_timestamp + timedelta(milliseconds=1)  # Add +1 ms for inclusive timestamp
+                elif len(logs) == 0 and since < timezone.now() - timedelta(hours=24):
+                    # If no logs where retrieved during the last 24hours,
+                    # move forward 1h to prevent stagnate ad vitam eternam
+                    self.last_collected_timestamps[f"perception_point_x_ray_{log_type}"] = since + timedelta(hours=1)
+                else:
+                    self.last_collected_timestamps[f"perception_point_x_ray_{log_type}"] = to + timedelta(milliseconds=1)  # Add +1 ms for inclusive timestamp
```

Comment:
Don't you mean something like:

```suggestion
                if logs:
                    if has_more_logs:
                        max_timestamp = max(int(log.get("timestamp_epoch") * 10000000) for log in logs)
                        last_timestamp = datetime.fromtimestamp(max_timestamp/10000000, tz=timezone.utc)
                        self.last_collected_timestamps[f"perception_point_x_ray_{log_type}"] = last_timestamp + timedelta(milliseconds=1)  # Add +1 ms for inclusive timestamp
                    else:
                        self.last_collected_timestamps[f"perception_point_x_ray_{log_type}"] = to + timedelta(milliseconds=1)  # Add +1 ms for inclusive timestamp
                elif since < timezone.now() - timedelta(hours=24):
                    # If no logs where retrieved during the last 24hours,
                    # move forward 1h to prevent stagnate ad vitam eternam
                    self.last_collected_timestamps[f"perception_point_x_ray_{log_type}"] = since + timedelta(hours=1)
```

Reply:
Your proposition add more complexity with nested 'if' conditions. I think mine is more clear and readable.. :p

Reply:
yes, but mine takes into account the early termination of `get_logs()` with the `evt_stop` being set, without updating the timestamps